data/trie/patriciaMerkleTrie.go

```diff
@@ -445,37 +449,71 @@ func (tr *patriciaMerkleTrie) ResetOldHashes() [][]byte {
 	return oldHashes
 }
 
-// Snapshot creates a new database in which the current state of the trie is saved.
-// If the maximum number of snapshots has been reached, the oldest snapshot is removed.
-func (tr *patriciaMerkleTrie) Snapshot() error {
+// Checkpoint adds the current state of the trie to the snapshot database
+func (tr *patriciaMerkleTrie) Checkpoint() error {
 	tr.mutOperation.Lock()
 	defer tr.mutOperation.Unlock()
 
+	if tr.root.isDirty() {
+		return ErrTrieNotCommited
+	}
+
 	if tr.snapshotInProgress {
 		return ErrSnapshotInProgress
 	}
 	tr.snapshotInProgress = true
 
-	err := tr.root.commit(false, 0, tr.db, tr.db)
+	if len(tr.snapshots) == 0 {
+		_, err := tr.newSnapshotDb()
+		if err != nil {
+			tr.snapshotInProgress = false
+			return err
+		}
+	}
+
+	lastSnapshotId := len(tr.snapshots) - 1
+	err := tr.root.commit(true, 0, tr.db, tr.snapshots[lastSnapshotId])
 	if err != nil {
+		tr.snapshotInProgress = false
 		return err
 	}
 
+	keys := tr.pruningBuffer
+	tr.pruningBuffer = make([][]byte, 0)
+	tr.snapshotInProgress = false
+
+	for i := range keys {
+		err = tr.removeFromDb(keys[i])
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// Snapshot creates a new database in which the current state of the trie is saved.
+// If the maximum number of snapshots has been reached, the oldest snapshot is removed.
+func (tr *patriciaMerkleTrie) Snapshot() error {
```

Comment:
This is not called at the current moment. An integration PR will follow?

Reply:
Yes. Snapshot will be called at the beginning of a new epoch, so the endOfEpoch is needed in order to call this function. Also for the checkpoint, an integration PR will follow.